File: machine-evolved-trainer/CreatureStructure.py
# This class must exactly match the CreatureStructure c++ class in MachineWorker.

from collections import namedtuple
import json
import uuid
from pprint import pprint

CAPSULE = namedtuple("CAPSULE", "id innerHeight radius positionX positionY positionZ quaternionX quaternionY quaternionZ quaternionW constraint")

class CreatureStructure:
	def __init__(self, jsonData = None):
		self.capsules = []
		self.numFeedbacks = 0
		self.inputs = {}
		
		if jsonData != None:
			self.buildFromJson(jsonData)

	def setInputs(self, inputs):
		self.inputs = inputs

	def setNumFeedbacks(self, num):
		self.numFeedbacks = num

	def setOscillatorStart(self, startRatio):
		self.oscillatorStart = startRatio

	def setOscillatorMultiplier(self, multiplier):
		self.oscillatorMultiplier = multiplier

	def setOscillatorCount(self, count):
		self.oscillatorCount = count
		
	def buildFromJson(self, jsonData):
		self.numFeedbacks = jsonData["feedbacks"]
		self.oscillatorStart = jsonData["oscillators"]["start"]
		self.oscillatorMultiplier = jsonData["oscillators"]["multiplier"]
		self.oscillatorCount = jsonData["oscillators"]["count"]
		self.inputs = jsonData["inputs"]

		for item in jsonData["capsules"]:
			capsule = CAPSULE(item["id"], item["innerHeight"], item["radius"], item["positionX"], item["positionY"], item["positionZ"], item["quaternionX"], item["quaternionY"], item["quaternionZ"], item["quaternionW"], item["constraint"])
			self.addCapsule(capsule)

	def addCapsule(self, capsule):
		self.capsules.append(capsule)

	# configJson: A dictonary with configuration details for the constraint. As of 17/9 in structure/generator/motors.
	def addCapsuleWithConstraint(self, innerHeight, radius, parentCapsule, configJson):
		# inner functions are quaternion linear algebra helper functions to avoid importing external libraries
		def quaternion_conjugate(q):
			w, x, y, z = q
			return (w, -x, -y, -z)

		def quaternion_mult(q1, q2):
			w1, x1, y1, z1 = q1
			w2, x2, y2, z2 = q2
			w = w1 * w2 - x1 * x2 - y1 * y2 - z1 * z2
			x = w1 * x2 + x1 * w2 + y1 * z2 - z1 * y2
			y = w1 * y2 + y1 * w2 + z1 * x2 - x1 * z2
			z = w1 * z2 + z1 * w2 + x1 * y2 - y1 * x2
			return w, x, y, z

		def quaternion_vector_mult(q1, v1):
			q2 = (0.0,) + v1
			return quaternion_mult(quaternion_mult(q1, q2), quaternion_conjugate(q1))[1:]

		middlePosition = (0, 0, .5*parentCapsule.innerHeight + parentCapsule.radius + .5*innerHeight + radius)
		parentQuaternion = (parentCapsule.quaternionW, parentCapsule.quaternionX, parentCapsule.quaternionY, parentCapsule.quaternionZ)		# Using w,x,y,z notation - not x,y,z,w like I do in c++
		rotatedPosition = quaternion_vector_mult(parentQuaternion, middlePosition)
		newPosition = (rotatedPosition[0] + parentCapsule.positionX, rotatedPosition[1] + parentCapsule.positionY, rotatedPosition[2] + parentCapsule.positionZ)

		configWithParent = configJson.copy()
		configWithParent["parentId"] = parentCapsule.id
		capsule = CAPSULE(str(uuid.uuid4()), innerHeight, radius, newPosition[0], newPosition[1], newPosition[2], parentCapsule.quaternionX, parentCapsule.quaternionY, parentCapsule.quaternionZ, parentCapsule.quaternionW, configWithParent)
		self.addCapsule(capsule)
		return capsule

	def getCapsules(self):
		return self.capsules

	def getJson(self):
		capsules = []
		for capsule in self.capsules:
			capsuleDictionary = capsule._asdict()

			capsules.append(capsuleDictionary)

		oscillators = { "start": self.oscillatorStart, "multiplier": self.oscillatorMultiplier, "count": self.oscillatorCount }

		return { "capsules": capsules, "feedbacks": self.numFeedbacks, "oscillators": oscillators, "inputs": self.inputs }

	def serialize(self):
		return json.dumps(getJson())

	def getNumInputs(self):
		# Per creature:
		numPerCreature = 0
		creatureKeys = ["root-orientation-x", "root-orientation-y", "root-orientation-z", "root-orientation-w", "z-position", "velocity-x", "velocity-y", "velocity-z"]
		for key in creatureKeys:
			if self.inputs[key] == 1:
				numPerCreature += 1
		if self.inputs["oscillators"] == 1:
			numPerCreature += self.oscillatorCount

		# Per capsule:
		numPerCapsule = 0
		capsuleKeys = ["capsule-position-x", "capsule-position-y", "capsule-position-z", "capsule-velocity-x", "capsule-velocity-y", "capsule-velocity-z", "capsule-angular-velocity-x", "capsule-angular-velocity-y", "capsule-angular-velocity-z"]
		for key in capsuleKeys:
			if self.inputs[key] == 1:
				numPerCapsule += 1

		# For motors:
		numForMotors = 0
		for capsule in self.capsules:
			if capsule.constraint:
				if "x-rotation" in capsule.constraint and self.inputs["motor-angle-x"] == 1:
					numForMotors += 1
				if "y-rotation" in capsule.constraint and self.inputs["motor-angle-y"] == 1:
					numForMotors += 1
				if "z-rotation" in capsule.constraint and self.inputs["motor-angle-z"] == 1:
					numForMotors += 1

		# For feedbacks:
		numForFeedbacks = 0
		if self.inputs["feedbacks"] == 1:
			numForFeedbacks = self.numFeedbacks

		return numPerCreature + len(self.capsules) * numPerCapsule + numForMotors + numForFeedbacks

	def getNumOutputs(self):
		num = 0

		for capsule in self.capsules:
			if capsule.constraint:
				if "x-rotation" in capsule.constraint:
					num += 1
				if "y-rotation" in capsule.constraint:
					num += 1
				if "z-rotation" in capsule.constraint:
					num += 1

		num += self.numFeedbacks

		return num

	def getNumConstraints(self):
		num = 0

		for capsule in self.capsules:
			if capsule.constraint:
				num = num + 1

		return num

File: MachineWorker/MachineWorker/CreatureStructure.cpp
#include "CreatureStructure.h"

// Constructor that initializes from a serialization
CreatureStructure::CreatureStructure(pt::ptree serialized) {
	numFeedbacks = serialized.get<int>("feedbacks");
	
	oscillatorStart = serialized.get<float>("oscillators.start");
	oscillatorMultiplier = serialized.get<float>("oscillators.multiplier");
	oscillatorCount = serialized.get<int>("oscillators.count");
	
	inputs.capsuleAngularVelocityX = serialized.get<int>("inputs.capsule-angular-velocity-x") == 1;
	inputs.capsuleAngularVelocityY = serialized.get<int>("inputs.capsule-angular-velocity-y") == 1;
	inputs.capsuleAngularVelocityZ = serialized.get<int>("inputs.capsule-angular-velocity-z") == 1;
	inputs.capsulePositionX = serialized.get<int>("inputs.capsule-position-x") == 1;
	inputs.capsulePositionY = serialized.get<int>("inputs.capsule-position-y") == 1;
	inputs.capsulePositionZ = serialized.get<int>("inputs.capsule-position-z") == 1;
	inputs.capsuleVelocityX = serialized.get<int>("inputs.capsule-velocity-x") == 1;
	inputs.capsuleVelocityY = serialized.get<int>("inputs.capsule-velocity-y") == 1;
	inputs.capsuleVelocityZ = serialized.get<int>("inputs.capsule-velocity-z") == 1;
	inputs.feedbacks = serialized.get<int>("inputs.feedbacks") == 1;
	inputs.motorAngleX = serialized.get<int>("inputs.motor-angle-x") == 1;
	inputs.motorAngleY = serialized.get<int>("inputs.motor-angle-y") == 1;
	inputs.motorAngleZ = serialized.get<int>("inputs.motor-angle-z") == 1;
	inputs.oscillators = serialized.get<int>("inputs.oscillators") == 1;
	inputs.rootOrientationX = serialized.get<int>("inputs.root-orientation-x") == 1;
	inputs.rootOrientationY = serialized.get<int>("inputs.root-orientation-y") == 1;
	inputs.rootOrientationZ = serialized.get<int>("inputs.root-orientation-z") == 1;
	inputs.rootOrientationW = serialized.get<int>("inputs.root-orientation-w") == 1;
	inputs.velocityX = serialized.get<int>("inputs.velocity-x") == 1;
	inputs.velocityY = serialized.get<int>("inputs.velocity-y") == 1;
	inputs.velocityZ = serialized.get<int>("inputs.velocity-z") == 1;
	inputs.zPosition = serialized.get<int>("inputs.z-position") == 1;

	for (pt::ptree::value_type &capsule : serialized.get_child("capsules")) {
		auto item = capsule.second;
		auto capsule = new CAPSULE();

		capsule->id = item.get<std::string>("id");
		capsule->innerHeight = item.get<double>("innerHeight");
		capsule->radius = item.get<double>("radius");
		capsule->positionX = item.get<double>("positionX");
		capsule->positionY = item.get<double>("positionY");
		capsule->positionZ = item.get<double>("positionZ");
		capsule->quaternionX = item.get<double>("quaternionX");
		capsule->quaternionY = item.get<double>("quaternionY");
		capsule->quaternionZ = item.get<double>("quaternionZ");
		capsule->quaternionW = item.get<double>("quaternionW");
		capsule->constraint = item.get_child("constraint");

		addCapsule(capsule);
	}
}

CreatureStructure::~CreatureStructure()
{
	for (CAPSULE* capsule : capsules)
		delete capsule;
}

int CreatureStructure::getNumInputs() {
	// Per creature:
	int numPerCreature = 0;
	numPerCreature += inputs.rootOrientationX ? 1 : 0;
	numPerCreature += inputs.rootOrientationY ? 1 : 0;
	numPerCreature += inputs.rootOrientationZ ? 1 : 0;
	numPerCreature += inputs.rootOrientationW ? 1 : 0;
	numPerCreature += inputs.zPosition ? 1 : 0;
	numPerCreature += inputs.velocityX ? 1 : 0;
	numPerCreature += inputs.velocityY ? 1 : 0;
	numPerCreature += inputs.velocityZ ? 1 : 0;
	numPerCreature += inputs.oscillators ? oscillatorCount : 0;

	// Per capsule:
	int numPerCapsule = 0;
	numPerCapsule += inputs.capsulePositionX ? 1 : 0;
	numPerCapsule += inputs.capsulePositionY ? 1 : 0;
	numPerCapsule += inputs.capsulePositionZ ? 1 : 0;
	numPerCapsule += inputs.capsuleVelocityX ? 1 : 0;
	numPerCapsule += inputs.capsuleVelocityY ? 1 : 0;
	numPerCapsule += inputs.capsuleVelocityZ ? 1 : 0;
	numPerCapsule += inputs.capsuleAngularVelocityX ? 1 : 0;
	numPerCapsule += inputs.capsuleAngularVelocityY ? 1 : 0;
	numPerCapsule += inputs.capsuleAngularVelocityZ ? 1 : 0;

	// For motors:
	int numForMotors = 0;
	for (auto capsule : capsules) {
		if (!capsule->constraint.empty()) {
			if (capsule->constraint.count("x-rotation") == 1 && inputs.motorAngleX)
				numForMotors++;
			if (capsule->constraint.count("y-rotation") == 1 && inputs.motorAngleY)
				numForMotors++;
			if (capsule->constraint.count("z-rotation") == 1 && inputs.motorAngleZ)
				numForMotors++;
		}
	}

	// For feedbacks:
	int numForFeedbacks = inputs.feedbacks ? numFeedbacks : 0;

	return numPerCreature + capsules.size() * numPerCapsule + numForMotors + numForFeedbacks;
}

int CreatureStructure::getNumOutputs() {
	int num = 0;

	for (auto capsule : capsules) {
		if (!capsule->constraint.empty()) {
			if (capsule->constraint.count("x-rotation") == 1)
				num++;
			if (capsule->constraint.count("y-rotation") == 1)
				num++;
			if (capsule->constraint.count("z-rotation") == 1)
				num++;
		}
	}

	num += numFeedbacks;

	return num;
}

std::vector<CreatureStructure::CAPSULE*> CreatureStructure::getCapsules() {
	return capsules;
}

CreatureStructure::CAPSULE* CreatureStructure::getCapsule(std::string id) {
	for (CAPSULE* capsule : capsules) {		// TODO: For optimalization, change std::vector into std::map or similar.
		if (capsule->id == id)
			return capsule;
	}
	return NULL;	// not found
}

void CreatureStructure::addCapsule(CAPSULE* capsule) {
	capsules.push_back(capsule);
}

